### ISSUE TITLE
The error prone sadly crashes my compiler

### ISSUE THREAD

**Issue by pffrandsen** [google/error-prone#547]
Code in Question:
        Assert.assertTrue(first.equals(first));
        Assert.assertFalse(first.equals("foo")); ---- Fails here
        Assert.assertFalse(first.equals(null));

 error-prone version: 2.0.17
     Stack Trace:
     java.lang.NullPointerException
        at com.google.errorprone.util.ASTHelpers.findMatchingMethods(ASTHelpers.java:473)
        at com.google.errorprone.bugpatterns.EqualsIncompatibleType.matchMethodInvocation(EqualsIncompatibleType.java:120)
        at com.google.errorprone.scanner.ErrorProneScanner.visitMethodInvocation(ErrorProneScanner.java:952)
        at com.google.errorprone.scanner.ErrorProneScanner.visitMethodInvocation(ErrorProneScanner.java:144)
        at com.sun.tools.javac.tree.JCTree$JCMethodInvocation.accept(JCTree.java:1614)
        at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:71)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:89)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:43)
        at com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
        at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
        at com.sun.source.util.TreeScanner.visitMethodInvocation(TreeScanner.java:509)
        at com.google.errorprone.scanner.ErrorProneScanner.visitMethodInvocation(ErrorProneScanner.java:958)
        at com.google.errorprone.scanner.ErrorProneScanner.visitMethodInvocation(ErrorProneScanner.java:144)
        at com.sun.tools.javac.tree.JCTree$JCMethodInvocation.accept(JCTree.java:1614)
        at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:71)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:89)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:43)
        at com.sun.source.util.TreeScanner.visitExpressionStatement(TreeScanner.java:433)
        at com.google.errorprone.scanner.ErrorProneScanner.visitExpressionStatement(ErrorProneScanner.java:758)
        at com.google.errorprone.scanner.ErrorProneScanner.visitExpressionStatement(ErrorProneScanner.java:144)
        at com.sun.tools.javac.tree.JCTree$JCExpressionStatement.accept(JCTree.java:1424)
        at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:71)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:89)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:43)
        at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
        at com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
        at com.sun.source.util.TreeScanner.visitBlock(TreeScanner.java:248)
        at com.google.errorprone.scanner.ErrorProneScanner.visitBlock(ErrorProneScanner.java:575)
        at com.google.errorprone.scanner.ErrorProneScanner.visitBlock(ErrorProneScanner.java:144)
        at com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:996)
        at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:71)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:89)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:43)
        at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
        at com.sun.source.util.TreeScanner.visitMethod(TreeScanner.java:206)
        at com.google.errorprone.scanner.ErrorProneScanner.visitMethod(ErrorProneScanner.java:943)
        at com.google.errorprone.scanner.ErrorProneScanner.visitMethod(ErrorProneScanner.java:144)
        at com.sun.tools.javac.tree.JCTree$JCMethodDecl.accept(JCTree.java:868)
        at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:71)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:89)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:43)
        at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
        at com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
        at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
        at com.sun.source.util.TreeScanner.visitClass(TreeScanner.java:187)
        at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:635)
        at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:144)
        at com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:778)
        at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:71)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:89)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:43)
        at com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
        at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
        at com.sun.source.util.TreeScanner.visitCompilationUnit(TreeScanner.java:144)
        at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:650)
        at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:144)
        at com.sun.tools.javac.tree.JCTree$JCCompilationUnit.accept(JCTree.java:561)
        at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:52)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:68)
        at com.google.errorprone.scanner.ErrorProneScannerTransformer.apply(ErrorProneScannerTransformer.java:43)
        at com.google.errorprone.ErrorProneAnalyzer.finished(ErrorProneAnalyzer.java:145)
        at com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:120)
        at com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1277)
        at com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1226)
        at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:840)
        at com.sun.tools.javac.main.Main.compile(Main.java:257)

**Comment by cushon** [comment]
@nglorioso has been doing some work on this check.

@pffrandsen, do you have a self-contained example that reproduces the crash?

**Comment by pffrandsen** [comment]
Orignial file
package com.partisia.auction.model;

import org.hamcrest.core.Is;
import org.junit.Assert;

import java.lang.reflect.Field;

/**
 * Created by ebo on 19-01-2017.
 */
class IdTester {

    @SuppressWarnings({"EqualsWithItself", "EqualsBetweenInconvertibleTypes", "ObjectEqualsNull", "SelfEquals"})
    static <T extends BaseModel> void testEqualsAndHashcode(T first, T second) throws Exception {
        setId(first, null);
        setId(second, null);

        Assert.assertThat(first.hashCode(), Is.is(second.hashCode()));
        Assert.assertFalse(first.equals(second));

        setId(first, 1);
        Assert.assertTrue(first.equals(first));
Two version for the next line
        Assert.assertFalse(first.equals("foo)); <--- Failes
        Assert.assertFalse(first.equals(new Object())); <--- Works
        Assert.assertFalse(first.equals(null));
        Assert.assertFalse(first.equals(second));

        setId(second, 1);
        Assert.assertTrue(first.equals(second));
        Assert.assertThat(first.hashCode(), Is.is(second.hashCode()));
    }

    private static <T extends BaseModel> void setId(T auction, Integer value) throws NoSuchFieldException, IllegalAccessException {
        Field id = BaseModel.class.getDeclaredField("id");
        id.setAccessible(true);
        id.set(auction, value);
    }
}

**Comment by nglorioso** [comment]
Thanks for the repro case! We've found the issue and hope to have a fix out shortly.

TL;DR: We were looking for 'equals' methods on the type variable T, which being a type variable, doesn't actually have any members. (BaseModel might have an override, for example, and we want to make sure the argument might override BaseModel).
